File: go/linked_list/reverseList206.go
package main

import (
	. "wtf-leetcode/linked_list/def"
)

func reverseList2(head *ListNode) *ListNode {
	// recurse to last node
	if head == nil || head.Next == nil {
		return head
	}
	last := reverseList2(head.Next)
	head.Next.Next = head
	head.Next = nil
	return last
}

func reverseList(head *ListNode) *ListNode {
	//     prev -> 1 -> 2 -> 3 -> nil
	// nil <- 1 <- 2 <- 3
	var prev, curr *ListNode = nil, head
	for curr != nil {
		p := curr
		curr = curr.Next
		p.Next = prev
		prev = p
	}
	return prev
}

func main() {
	l1 := &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3}}}
	rel1 := reverseList(l1)
	PrintList(rel1)
}

File: python/listed_list/reverseList206.py
from lib.node import *


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        p = None
        p1 = head
        while p1:
            p2 = p1
            p1 = p1.next
            p2.next = p
            p = p2
        return p

if __name__ == '__main__':
    s = Solution()
    l = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    print_nodes(s.reverseList(l))
